topi/python/topi/cuda/conv2d_transpose_nchw.py

```diff
@@ -1,121 +1,199 @@
-        """Internal travserse function"""
-        # inline all one-to-one-mapping operators except the last stage (output)
-        if tag.is_injective(OP.tag):
-            if OP not in s.outputs:
-                s[OP].compute_inline()
-            for tensor in OP.input_tensors:
-                if tensor.op.input_tensors and tensor.op not in scheduled_ops:
-                    traverse(tensor.op)
-        # schedule conv2d_transpose_nchw
-        if 'conv2d_transpose_nchw' in OP.tag:
-            temp = OP.input_tensors[0]
-            DilatedInput = temp.op.input_tensors[0]
-            s[DilatedInput].compute_inline()
-            Filter = OP.input_tensors[1]
-            Output = OP.output(0)
-            schedule(temp, Filter, Output)
-
-        scheduled_ops.append(OP)
-
-    traverse(outs[0].op)
-    return s
+@autotvm.task.register_topi_compute(nn.conv2d_transpose_nchw, ['cuda', 'gpu'], "direct")
+def conv2d_transpose_nchw_cuda(cfg, Input, Filter, strides, padding, out_dtype):
+    """Transposed 2D convolution nchw forward operator.
 
+    Parameters
+    ----------
+    cfg: ConfigEntity
+        The config for this template
+    Input : tvm.Tensor
+        4-D with shape [batch, in_channel, in_height, in_width]
+    Filter : tvm.Tensor
+        4-D with shape [in_channel, num_filter, filter_height, filter_width]
+    strides : tuple of two ints
+        The spatial stride along height and width
+    padding : int or str
+        Padding size, or ['VALID', 'SAME']
+    out_dtype: str
+        The output type. This is used in mixed precision
 
-@generic.schedule_conv2d_transpose_nchw.register(["cuda", "gpu"])
-def schedule_conv2d_transpose_nchw(outs):
-    """Schedule for conv2d_transpose_nchw.
+    Returns
+    -------
+    Output : tvm.Tensor
+        4-D with shape [batch, out_channel, out_height, out_width]
+    """
+    batch, in_c, in_h, in_w = get_const_tuple(Input.shape)
+    _, out_c, filter_h, filter_w = get_const_tuple(Filter.shape)
+    stride_h, stride_w = strides
+
+    # attach stride info to config, this is used in schedule space definition
+    cfg.stride = strides
+
+    # padding stage
+    fpad_top, fpad_left, fpad_bottom, fpad_right = nn.get_pad_tuple(padding, (filter_h, filter_w))
+    bpad_top = filter_h - 1 - fpad_top
+    bpad_bottom = filter_h - 1 - fpad_bottom
+    bpad_left = filter_w - 1 - fpad_left
+    bpad_right = filter_w - 1 - fpad_right
+
+    # padding stage
+    FirstPad = nn.pad(Input,
+                      [0, 0, (bpad_top + stride_h - 1) // stride_h,
+                       (bpad_left + stride_w - 1) // stride_w],
+                      [0, 0, (bpad_bottom + stride_h - 1) // stride_h,
+                       (bpad_right + stride_w - 1) // stride_w], name='FirstPad')
+
+    # remove extra padding introduced by dilatation
+    border_h = (stride_h - bpad_top % stride_h) % stride_h
+    border_w = (stride_w - bpad_left % stride_w) % stride_w
+
+    # dilation stage
+    data = FirstPad
```

Comment:
Hi @merrymercy may I ask the question that why the computing sequence of Conv2DTranspose in CUDA is `pad -> dilate -> conv` while on other platforms are `dilate -> pad -> conv`? I have encountered building issue with `stride > 2` for CUDA - not root caused yet.

Reply:
I want to eliminate the multiplications of zeros introduced by dilations. This is required to get the best performance. So I have to do dilation after pad, so the zeros introduced by dilation can be passed to the later stage. Then, these multiplications of zeros will be eliminated by the simplifier in tvm.

I agree that this template will have issues with some input shapes. I do not handle all cases perfectly. Some fixes are required for both performance and correctness.

Dilation will introduce many multiplications of zeros. Currently, I use unroll + inline + simplifier in tvm to eliminate them. I will add a tutorial about this as suggested in https://github.com/dmlc/tvm/issues/1887.

Reply:
Got it. Thanks for your kind explanation.